simpeg/potential_fields/gravity/_numba_functions.py

```diff
@@ -163,6 +163,170 @@ def _sensitivity_gravity(
+    cell_nodes : (n_active_cells, 8) numpy.ndarray
+        Array of integers, where each row contains the indices of the nodes for
+        each active cell in the mesh.
+    kernel_func : callable
+        Kernel function that will be evaluated on each node of the mesh. Choose
+        one of the kernel functions in ``choclo.prism``.
+    constant_factor : float
+        Constant factor that will be used to multiply each element of the
+        sensitivity matrix.
+    vector : (n_receivers) numpy.ndarray
+        Array that represents the vector used in the dot product.
+    result : (n_active_cells) numpy.ndarray
+        Running result array where the output of the dot product will be added to.
+
+    Notes
+    -----
+    This function is meant to be run in parallel. This implementation splits
+    the task between threads on an outer loop. Each thread writes to their own
+    slice of a running result array to avoid race conditions. By the end, the
+    results obtained by each thread are added to obtain the final result.
+
+    """
+    n_receivers = receivers.shape[0]
+    n_nodes = nodes.shape[0]
+    n_cells = cell_nodes.shape[0]
+    n_threads = get_num_threads()
+    # Estimate number of receivers that will be assigned to each thread
+    chunk_size = n_receivers // n_threads
+    # Allocate result arrays, one per thread
+    result_per_thread = np.zeros((n_threads, n_cells), dtype=result.dtype)
+    # Split the job per thread
+    for thread_index in prange(n_threads):
+        # Allocate vector for kernels evaluated on mesh nodes.
+        kernels = np.empty(n_nodes)
+        # Evaluate kernel function on each node, for each receiver location in
+        # the assigned chunk to this thread.
+        start = thread_index * chunk_size
+        end = (
+            (thread_index + 1) * chunk_size
+            if thread_index < n_threads - 1
+            else n_receivers
+        )
+        for i in range(start, end):
+            for j in range(n_nodes):
+                kernels[j] = _evaluate_kernel(
+                    receivers[i, 0],
+                    receivers[i, 1],
+                    receivers[i, 2],
+                    nodes[j, 0],
+                    nodes[j, 1],
+                    nodes[j, 2],
+                    kernel_func,
+                )
+            # Compute sensitivity matrix elements for the i-th row, multiply them
+            # by the i-th element of the vector and add them to the running result
+            # array.
+            for k in range(n_cells):
+                result_per_thread[thread_index, k] += (
+                    constant_factor
+                    * vector[i]
+                    * kernels_in_nodes_to_cell(
+                        kernels,
+                        cell_nodes[k, :],
+                    )
+                )
+    # Reduce the operation and write into the result array
+    result[:] += np.sum(result_per_thread, axis=0)
```

Comment:
```suggestion
    n_receivers = receivers.shape[0]
    n_nodes = nodes.shape[0]
    n_cells = cell_nodes.shape[0]
    result[:] = 0
    # Evaluate kernel function on each node, for each receiver location
    for i in prange(n_receivers):
        # Allocate vector for kernels evaluated on mesh nodes
        kernels = np.empty(n_nodes)
        local_row = np.empty(n_cells)
        for j in range(n_nodes):
            kernels[j] = _evaluate_kernel(
                receivers[i, 0],
                receivers[i, 1],
                receivers[i, 2],
                nodes[j, 0],
                nodes[j, 1],
                nodes[j, 2],
                kernel_func,
            )
        # Compute fields from the kernel values
        for k in range(n_cells):
            local_row[k] += (
                constant_factor
                * vector[i]
                * kernels_in_nodes_to_cell(
                    kernels,
                    cell_nodes[k, :],
                )
            )
        result += local_row
```
This should get numba to pickup that result is a reduction operation, and cause it to handle it correctly in the prange

Reply:
Just did some local tests myself too, relying on `numba` to do the reduction produces the expected timings on my end (meaning the parallel forward and adjoint operations both had similar timings), what kind of array sizes were you experimenting with?

Here's my timings of a simple example: [the gist](https://gist.github.com/jcapriot/cdbd76e0f35013c60a8f02601303baf4)

Reply:
I think I was working with small arrays on Friday. I'm running better benchmarks now and I don't see any difference between the two implementations:

```python
import time
import numpy as np
import discretize
from simpeg.potential_fields import gravity
from simpeg import maps
from simpeg.utils import model_builder


def get_simulation(**kwargs):
    # Observation points
    x = np.linspace(-20.0, 20.0, 40)
    x, y = np.meshgrid(x, x)
    z = 5.0 * np.ones_like(x)
    coordinates = np.vstack((x.ravel(), y.ravel(), z.ravel())).T
    receivers = gravity.receivers.Point(coordinates, components="gz")
    source_field = gravity.sources.SourceField(receiver_list=[receivers])
    survey = gravity.survey.Survey(source_field)

    # Mesh
    dh = 5.0
    hx = [(dh, 60)]
    mesh = discretize.TensorMesh([hx, hx, hx], "CCN")

    # Define densities
    densities = 1e-10 * np.ones(mesh.n_cells)
    ind_sphere = model_builder.get_indices_sphere(
        np.r_[0.0, 0.0, -20.0], 10.0, mesh.cell_centers
    )
    densities[ind_sphere] = 0.2

    # Define simulation
    mapping = maps.IdentityMap(nP=mesh.n_cells)
    default_kwargs = {
        "store_sensitivities": "ram",
        "engine": "choclo",
    }
    default_kwargs.update(kwargs)
    simulation = gravity.simulation.Simulation3DIntegral(
        survey=survey,
        mesh=mesh,
        rhoMap=mapping,
        **default_kwargs,
    )
    return simulation, densities


if __name__ == "__main__":
    simulation, densities = get_simulation(store_sensitivities="forward_only")
    simulation.model = densities # assign model so rhoDeriv gets defined (see #1621)
    print(f"n_data: {simulation.survey.nD}, n_params: {simulation.mesh.n_cells}")

    # Run once to compile Numba functions
    vector = np.random.default_rng(seed=42).uniform(size=simulation.survey.nD)
    simulation.Jtvec(densities, vector)

    # Benchmark
    start = time.perf_counter()
    simulation.Jtvec(densities, vector)
    end = time.perf_counter()
    print(f"Elapsed time: {end - start:2e}s")
```

With 4e601b8fca3d47087753ac3797fc9eac0611686b:
```
n_data: 1600, n_params: 216000
Elapsed time: 3.510344e+00s
```

After applying your suggestions:
```
n_data: 1600, n_params: 216000
Elapsed time: 3.660166e+00s
```

The time difference between the two is within the error.

Let's apply your suggestions then, since the code ends up being simpler.